File: .github/scripts/jules-generate-agent-skills.py
import os
import sys
import asyncio

try:
    # Assuming the Python SDK exposes the client similarly
    from google_jules import jules
except ImportError as e:
    print(f"::error::Failed to dynamically import google_jules: {e}")
    sys.exit(1)

async def run():
    if not os.environ.get("JULES_API_KEY"):
        print("::error::JULES_API_KEY environment variable is missing.")
        sys.exit(1)

    # 1. Get context about the current repository from environment variables
    github_repository = os.environ.get("GITHUB_REPOSITORY", "")
    repo_parts = github_repository.split("/")
    owner = repo_parts[0] if len(repo_parts) > 0 and repo_parts[0] else "unknown"
    repo = repo_parts[1] if len(repo_parts) > 1 and repo_parts[1] else "unknown"
    ref = os.environ.get("GITHUB_REF", "refs/heads/main")

    base_branch = "main"
    if ref.startswith("refs/heads/"):
        base_branch = ref.replace("refs/heads/", "")

    print(f"Starting Jules session for {owner}/{repo} on branch {base_branch}")

    # 2. Construct the prompt for generating Agent Skills
    prompt = """Analyze this repository and suggest Agent Skills to improve automation of common or complex tasks.

Use the Agent Skills specification located at https://agentskills.io/specification.md as a reference for formatting and structuring the skills.

Tasks:
1. Review the repository structure, code, and existing workflows.
2. Identify 1 to 3 areas where an Agent Skill could be beneficial (e.g., code review, automated testing, boilerplate generation, or specific formatting rules).
3. Create the corresponding Agent Skills configuration files (e.g., in a `.jules/skills` directory or similar, as per the specification).
4. Provide a brief explanation of what each skill does and why it is useful for this repository."""

    # 3. Create a new Jules session
    session = await jules.session(
        prompt=prompt,
        source={
            "github": f"{owner}/{repo}",
            "baseBranch": base_branch
        },
        autoPr=True
    )

    print(f"Session created successfully. ID: {session.id}")

    # 4. Monitor the progress
    async for activity in session.stream():
        if activity.type == "planGenerated":
            print(f"[Plan Generated] {len(activity.plan.steps)} steps.")
        elif activity.type == "progressUpdated":
            print(f"[Progress Updated] {activity.title}")
        elif activity.type == "sessionCompleted":
            print("[Session Completed]")

    # 5. Wait for the final outcome
    outcome = await session.result()

    if outcome.state == "failed":
        print("::error::Session failed.")
        sys.exit(1)

    print("Session finished successfully.")

    if getattr(outcome, "pullRequest", None):
        print(f"Pull Request created: {outcome.pullRequest.url}")

if __name__ == "__main__":
    asyncio.run(run())

File: .github/scripts/jules-generate-agent-skills.mjs
import * as core from "@actions/core";
import * as github from "@actions/github";

let julesModule;
try {
  // Use dynamic import to safely resolve ESM packages and bypass strict CommonJS require() limits
  julesModule = await import("@google/jules-sdk");
} catch (err) {
  core.setFailed(`Failed to dynamically import @google/jules-sdk: ${err.message}`);
  process.exit(1);
}

const jules = julesModule.jules || julesModule.default?.jules || julesModule.default;

if (!jules) {
  core.setFailed("Could not extract the jules object from the SDK exports.");
  process.exit(1);
}

if (!process.env.JULES_API_KEY) {
  core.setFailed("JULES_API_KEY environment variable is missing.");
  process.exit(1);
}

const context = github.context;
const owner = context.repo.owner || process.env.GITHUB_REPOSITORY?.split("/")[0] || "unknown";
const repo = context.repo.repo || process.env.GITHUB_REPOSITORY?.split("/")[1] || "unknown";
const ref = context.ref || process.env.GITHUB_REF || "refs/heads/main";

let baseBranch = "main";

jules.configure({ apiKey: process.env.JULES_API_KEY });

if (ref.startsWith("refs/heads/")) baseBranch = ref.replace("refs/heads/", "");

core.info(`Starting Jules session for ${owner}/${repo} on branch ${baseBranch}`);

const prompt = `Analyze this repository and suggest Agent Skills to improve automation of common or complex tasks.

Use the Agent Skills specification located at https://agentskills.io/specification.md as a reference for formatting and structuring the skills.

Tasks:
1. Review the repository structure, code, and existing workflows.
2. Identify 1 to 3 areas where an Agent Skill could be beneficial (e.g., code review, automated testing, boilerplate generation, or specific formatting rules).
3. Create the corresponding Agent Skills configuration files (e.g., in a \`.jules/skills\` directory or similar, as per the specification).
4. Provide a brief explanation of what each skill does and why it is useful for this repository.`;

const session = await jules.session({
  apiKey: process.env.JULES_API_KEY,
  prompt,
  source: { github: `${owner}/${repo}`, baseBranch },
  autoPr: true,
});

core.info(`Session created successfully. ID: ${session.id}`);

for await (const activity of session.stream()) {
  if (activity.type === "planGenerated") core.info(`[Plan Generated] ${activity.plan.steps.length} steps.`);
  if (activity.type === "progressUpdated") core.info(`[Progress Updated] ${activity.title}`);
  if (activity.type === "sessionCompleted") core.info(`[Session Completed]`);
}

const outcome = await session.result();
if (outcome.state === "failed") {
  core.setFailed("Session failed.");
  process.exit(1);
}

core.info("Session finished successfully.");
if (outcome.pullRequest) {
  core.info(`Pull Request created: ${outcome.pullRequest.url}`);
}
